### ISSUE TITLE
SDK is causing our app to crash in a second location 

### ISSUE THREAD

**Issue by alexjheu** [SurveyMonkey/surveymonkey-android-sdk#55]
Our Android App crashes while trying to unsubscribe from a plan. This is when we have a survey pop-up when the user taps the "unsubscribe" button. 

Our crash log points out that the crash happened in the survey monkey sdk. . It looks like your SDK code SMFeedbackFragment.onPageFinished() calls SMFeedbackFragment.this.mProgressDialog.dismiss() even if it's already dismissed.

**Comment by ayeung** [comment]
@alexjheu Would you mind posting the stack trace here?

**Comment by alexjheu** [comment]
@ayeung 
--------- beginning of crash
08-22 16:48:08.822 4846 4846 E AndroidRuntime: java.lang.IllegalArgumentException: View=DecorView@f08adb4[SM_SurveyMonkey] not attached to window manager**
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.view.WindowManagerGlobal.findViewLocked(WindowManagerGlobal.java:503)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.view.WindowManagerGlobal.removeView(WindowManagerGlobal.java:412)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.view.WindowManagerImpl.removeViewImmediate(WindowManagerImpl.java:136)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.app.Dialog.dismissDialog(Dialog.java:599)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.app.Dialog.dismiss(Dialog.java:582)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at com.surveymonkey.surveymonkeyandroidsdk.SMFeedbackFragment$a.onPageFinished(SMFeedbackFragment.java:287)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at xf.c(SourceFile:251)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at afs.handleMessage(SourceFile:72)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.os.Handler.dispatchMessage(Handler.java:102)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.os.Looper.loop(Looper.java:154)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at android.app.ActivityThread.main(ActivityThread.java:6781)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at java.lang.reflect.Method.invoke(Native Method)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
08-22 16:48:08.822 4846 4846 E AndroidRuntime: at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)

**Comment by alexjheu** [comment]
Thanks. Let me know if there are any updates.

**Comment by alexjheu** [comment]
@ayeung any update on this?

**Comment by alexjheu** [comment]
@ayeung This issue has been reproduced today and devices Samsung S7, Android 8.0 and Samsung S8 Active, Android 7.0. 
![vvmtest_voicemailtranscribedwheninplan14testunsubscribe_2018-09-26t10-05-36-365](https://user-images.githubusercontent.com/40275429/46094600-5de80580-c188-11e8-8465-5c6b7dd4e1c2.png)

**Comment by ayeung** [comment]
Looking into this more, I don't think calling dismiss when it's already dismissed will cause any errors.  If you look at the implementation dismiss() in Android, it actually checks if it's showing before doing anything.  It's likely that the activity that the fragment is attached to was finished, which would cause the same error.  Here's a thread of someone having the same error: https://stackoverflow.com/questions/22924825/view-not-attached-to-window-manager-crash

I would make sure that you keep that activity alive that launched the surveymonkey fragment.

**Comment by cyn-ph** [comment]
@ayeung do have any update on this?

**Comment by ayeung** [comment]
@cyn-ph As mentioned in the response above, this is most likely not an issue in the SDK.  The caller needs to ensure that the calling activity is still alive when launching the fragment.

**Comment by cyn-ph** [comment]
@ayeung how am I suppose to ensure that the calling activity is still alive if the user presses home button? 
Look at this flow: 

- we start the activity and set the `SMFeedbackFragment` 
- `SMFeedbackFragment`'s `onCreate()` starts a `RetrieveSPageTask` 
- User presses home button and task hasn't finished, system kills the app
- `RetrieveSPageTask` finishes and inside its `onPostExecute()` there is a call to `loadSurveyPage()`, inside this method you set a listener ` mWebView.setWebViewClient(new SMWebviewClient());`
- When the callback `onPageFinished()` from the above listener gets called this line is executed `mProgressDialog.dismiss();` which is the one generating the crash, becase the activity is not in memory anymore and you're not cancelling the AsyncTask.

```   
 java.lang.IllegalArgumentException: View=DecorView@a8545f2[] not attached to window manager
        at android.view.WindowManagerGlobal.findViewLocked(WindowManagerGlobal.java:497)
        at android.view.WindowManagerGlobal.removeView(WindowManagerGlobal.java:404)
        at android.view.WindowManagerImpl.removeViewImmediate(WindowManagerImpl.java:124)
        at android.app.Dialog.dismissDialog(Dialog.java:367)
        at android.app.Dialog.dismiss(Dialog.java:346)
        at com.surveymonkey.surveymonkeyandroidsdk.SMFeedbackFragment$SMWebviewClient.onPageFinished(SMFeedbackFragment.java:287)
        at xy.c(SourceFile:228)
        at axy.handleMessage(SourceFile:72)
        at android.os.Handler.dispatchMessage(Handler.java:110)
        at android.os.Looper.loop(Looper.java:203)
        at android.app.ActivityThread.main(ActivityThread.java:6251)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1075)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
```

**Comment by ayeung** [comment]
Thanks for the info @cyn-ph , Will take a look at this as a part of the next release.

**Comment by ayeung** [comment]
Will be fixed in next release: 1.0.7

**Comment by cyn-ph** [comment]
Perfect! One more question @ayeung,  when is this release suppose to happen?

**Comment by ayeung** [comment]
Within the next week or so

**Comment by ayeung** [comment]
Just released 2.0.0.  Check it out when you get the chance.  I'll be closing this issue for now, feel free to reopen and/or comment if you are still having issues.

Note that I did run into a situation where the previous version of the library was cached on Android Studio, so I would also invalidate the cache to make sure you are grabbing the latest if you run into any issues.
